python/xir/program.py

```diff
@@ -637,6 +648,143 @@ def merge(*programs) -> "XIRProgram":
 
         return result
 
+    @staticmethod
+    def resolve(library: Mapping[str, "XIRProgram"], name: str) -> "XIRProgram":
+        """Resolves the includes of an XIR program using an XIR library.
+
+        Args:
+            library (Mapping[str, XIRProgram]): mapping from names to XIR programs
+            name (str): name of the root XIR program to resolve
+
+        Returns:
+            XIRProgram: XIR program obtained by recursively merging each XIR
+            program starting from the root XIR program
+
+        **Example**
+
+        Consider an XIR program ``main`` which includes another XIR program ``util``:
+
+        .. code-block:: python
+
+            import xir
+
+            # Create two simple XIR programs.
+            main_program = xir.parse_script("use util; H2 | [0, 1];")
+            util_program = xir.parse_script("gate H2, 0, 2; gate H2: H | [0]; H | [1]; end;")
+
+        The ``main`` XIR program can be resolved using
+
+        .. code-block:: python
+
+            # Define a library containing the XIR programs in the resolution scope.
+            library = {"main": main_program, "util": util_program}
+
+            # Resolve the imports from the main XIR program.
+            resolved_main_program = xir.XIRProgram.resolve(library, "main")
+
+            # Display the result.
+            print(resolved_main_program.serialize())
+
+        The output is
+
+        .. code-block:: none
+
+            gate H2, 0, 2;
+
+            gate H2:
+                H | [0];
+                H | [1];
+            end;
+
+            H2 | [0, 1];
+        """
+        resolved_names = XIRProgram._resolve_include_order(library, name, stack=set(), cache={})
+        resolved_programs = map(library.get, resolved_names)
+
+        resolved_program = XIRProgram.merge(*resolved_programs)
+        resolved_program._includes = []  # pylint: disable=protected-access
```

Comment:
Hmm, I'm having some trouble coming up with a situation where setting `include_includes=True` is useful.  One reason is that it can already be done in terms of the current implementation.  Specifically,
```python3
resolved_program = XIRProgram(library, name, include_includes=True)
```
is equivalent to
```python3
includes = library[name].includes

resolved_program = XIRProgram(library, name)

for include in includes:
    resolved_program.add_include(include)
```
The other issue with `include_includes=True` is that resolving a resolved program will almost certainly emit a bunch of warnings due to duplicate declarations and definitions.

Reply:
I'm not sure I understand. Currently, `XIRProgram.merge` will return a program in which the new `resolved_program` will have all includes from the input programs (so `include_includes=True` as it's written now). That's why we need to have the `resolved_program._includes = []` line. 

My suggestion was to simply be able to move that specific line into `XIRProgram.merge` so that it can be done without having to access a private attribute externally (e.g. by `self._includes = []`). 

Does this make sense?

Reply:
> My suggestion was to simply be able to move that specific line into `XIRProgram.merge` so that it can be done without having to access a private attribute externally (e.g. by `self._includes = []`).

Oh, sorry! I thought you were talking about adding an argument to `XIRProgram.resolve()` (not `XIRProgram.merge()`). 🤦‍♂️

In that case, I see your point although I'm not sure adding such a parameter is the best solution since it couples the merging routine with the include-clearing mechanism.  Instead, what do you think about adding a function such as
```python
XIRProgram.clear_includes()
```
Now, the caller can choose to clear the includes whenever they please.

Reply:
No worries! That was basically my other suggestion, so that sounds great. 🙂 👍

Reply:
Done: `XIRProgram` can now be used as follows:
```python
import xir

program = xir.parse_script("use xstd;")
program.clear_includes()

assert program.serialize() == ""
```
I think we can revisit adding "clear" functions for the other attributes when a use case for them is known.